File: airs_experiment/python/run_02_cfa.py
#!/usr/bin/env python3
"""
AIRS Phase 2: Confirmatory Factor Analysis (CFA)
Refactored from 02_CFA_Experiment.ipynb

Validates EFA-derived factor structure using independent holdout sample.
Method: semopy CFA | Fit criteria: CFI/TLI >= 0.95, RMSEA <= 0.08
Outputs: tables/cfa_summary.json, factor loadings, CR/AVE metrics
"""

import sys
import json
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import pingouin as pg
from semopy import Model
from semopy.stats import calc_stats

warnings.filterwarnings('ignore')

# ═══════════════════════════════════════════════════════════════════════════════
# CONFIGURATION
# ═══════════════════════════════════════════════════════════════════════════════

POPULATION = "combined"

CONSTRUCT_NAME_MAP = {
    'Factor1': 'PerfExp', 'Factor2': 'EffortExp', 'Factor3': 'SocialInf',
    'Factor4': 'FacilCond', 'Factor5': 'HedonicMot', 'Factor6': 'PriceValue',
    'Factor7': 'Habit', 'Factor8': 'AITrust'
}

DROPPED_ITEMS = {
    'Voluntariness': ['VO1', 'VO2'],
    'Explainability': ['EX1', 'EX2'],
    'EthicalRisk': ['ER1', 'ER2'],
    'Anxiety': ['AX1', 'AX2']
}


def build_cfa_syntax(structure):
    """Build semopy CFA model syntax from factor structure dict."""
    lines = []
    for factor, items in structure.items():
        lines.append(f"{factor} =~ {' + '.join(items)}")
    return '\n'.join(lines)


def compute_srmr(model, df):
    """Manually compute SRMR from a fitted semopy model."""
    sigma = model.calc_sigma()[0]
    observed_cov = df.cov().values
    n_vars = observed_cov.shape[0]

    # Convert model-implied covariance to correlation
    d_sigma = np.sqrt(np.diag(sigma))
    sigma_corr = sigma / np.outer(d_sigma, d_sigma)

    # Convert observed covariance to correlation
    d_obs = np.sqrt(np.diag(observed_cov))
    obs_corr = observed_cov / np.outer(d_obs, d_obs)

    # SRMR = sqrt(mean of squared residual correlations in lower triangle)
    residuals = obs_corr - sigma_corr
    lower_tri = residuals[np.tril_indices(n_vars, k=-1)]
    srmr = np.sqrt(np.mean(lower_tri**2))
    return srmr


def main():
    print("=" * 90)
    print("PHASE 2: CONFIRMATORY FACTOR ANALYSIS (CFA)")
    print("=" * 90)

    # ── Load Data ─────────────────────────────────────────────────────────
    df_full = pd.read_csv('data/AIRS_holdout.csv')
    print(f"\nLoaded holdout data: {df_full.shape[0]} rows x {df_full.shape[1]} columns")

    with open('data/experiment_config.json', 'r') as f:
        config = json.load(f)

    # ── Load EFA Results ──────────────────────────────────────────────────
    with open('tables/efa_summary.json', 'r') as f:
        efa_summary = json.load(f)

    # Build CFA structure from EFA results
    CFA_STRUCTURE = {}
    factor_struct = efa_summary.get('validated_factor_structure', {})
    for factor_name, items in factor_struct.items():
        mapped = CONSTRUCT_NAME_MAP.get(factor_name, factor_name)
        CFA_STRUCTURE[mapped] = items

    print(f"\nCFA Factor Structure ({len(CFA_STRUCTURE)} factors):")
    for factor, items in CFA_STRUCTURE.items():
        print(f"  {factor}: {items}")

    # ── Load Item Metadata ────────────────────────────────────────────────
    metadata = {}
    item_json_path = Path('../data/airs_28item_complete.json')
    if item_json_path.exists():
        with open(item_json_path, 'r') as f:
            item_data = json.load(f)
        items_raw = item_data.get('items', [])
        if isinstance(items_raw, list) and len(items_raw) > 0 and isinstance(items_raw[0], dict):
            for item in items_raw:
                metadata[item['item_id']] = item

    # ── Prepare CFA Data ─────────────────────────────────────────────────
    all_cfa_items = [item for items in CFA_STRUCTURE.values() for item in items]
    df_cfa = df_full[all_cfa_items].copy()
    print(f"\nCFA data prepared: {df_cfa.shape[0]} observations x {df_cfa.shape[1]} items")

    # ── Build & Fit CFA Model ─────────────────────────────────────────────
    model_syntax = build_cfa_syntax(CFA_STRUCTURE)
    print(f"\nCFA Model Syntax:\n{model_syntax}\n")

    cfa_model = Model(model_syntax)
    cfa_model.fit(df_cfa)
    fit_success = True

    # ── Fit Statistics ────────────────────────────────────────────────────
    stats_df = calc_stats(cfa_model)

    chi2 = stats_df.loc['Value', 'chi2'] if 'chi2' in stats_df.columns else None
    df_val = stats_df.loc['Value', 'DoF'] if 'DoF' in stats_df.columns else None
    cfi = stats_df.loc['Value', 'CFI'] if 'CFI' in stats_df.columns else None
    tli = stats_df.loc['Value', 'TLI'] if 'TLI' in stats_df.columns else None
    rmsea = stats_df.loc['Value', 'RMSEA'] if 'RMSEA' in stats_df.columns else None
    aic = stats_df.loc['Value', 'AIC'] if 'AIC' in stats_df.columns else None
    bic = stats_df.loc['Value', 'BIC'] if 'BIC' in stats_df.columns else None

    print("=" * 90)
    print("CFA FIT STATISTICS")
    print("=" * 90)

    fit_items = [
        ('χ²', chi2, None, None),
        ('df', df_val, None, None),
        ('CFI', cfi, 0.95, 'higher'),
        ('TLI', tli, 0.95, 'higher'),
        ('RMSEA', rmsea, 0.08, 'lower'),
        ('AIC', aic, None, None),
        ('BIC', bic, None, None),
    ]
    for name, val, threshold, direction in fit_items:
        if val is not None:
            status = ''
            if threshold:
                if direction == 'higher':
                    status = ' ✓' if val >= threshold else ' ⚠️'
                else:
                    status = ' ✓' if val <= threshold else ' ⚠️'
            print(f"  {name}: {val:.3f}{status}")

    # ── SRMR (Manual) ────────────────────────────────────────────────────
    srmr_value = compute_srmr(cfa_model, df_cfa)
    srmr_status = '✓' if srmr_value <= 0.08 else '⚠️'
    print(f"  SRMR: {srmr_value:.4f} {srmr_status}")

    # ── Factor Loadings ───────────────────────────────────────────────────
    params = cfa_model.inspect(std_est=True)
    loadings = params[params['op'] == '~']
    loading_records = []
    for _, row in loadings.iterrows():
        loading_records.append({
            'Factor': row['lval'],
            'Item': row['rval'],
            'Loading': row['Est. Std'],
            'SE': row['Std. Err'],
        })
    loading_df = pd.DataFrame(loading_records)

    print("\n" + "=" * 90)
    print("FACTOR LOADINGS (Standardized)")
    print("=" * 90)
    for factor in sorted(CFA_STRUCTURE.keys()):
        print(f"\n  {factor}:")
        frows = loading_df[loading_df['Factor'] == factor]
        for _, r in frows.iterrows():
            quality = '✓' if r['Loading'] >= 0.70 else ('○' if r['Loading'] >= 0.50 else '⚠️')
            print(f"    {r['Item']}: {r['Loading']:.3f} (SE={r['SE']:.3f}) {quality}")

    # Loading quality summary
    excellent = len(loading_df[loading_df['Loading'] >= 0.70])
    strong = len(loading_df[(loading_df['Loading'] >= 0.50) & (loading_df['Loading'] < 0.70)])
    acceptable = len(loading_df[(loading_df['Loading'] >= 0.40) & (loading_df['Loading'] < 0.50)])
    weak = len(loading_df[loading_df['Loading'] < 0.40])
    print(f"\n  Summary: Excellent(≥.70)={excellent}, Strong(.50-.70)={strong}, Acceptable(.40-.50)={acceptable}, Weak(<.40)={weak}")

    # ── Factor Correlations ───────────────────────────────────────────────
    factors = sorted(CFA_STRUCTURE.keys())
    cov_params = params[(params['op'] == '~~') & (params['lval'] != params['rval'])]

    corr_matrix = pd.DataFrame(1.0, index=factors, columns=factors)
    for _, row in cov_params.iterrows():
        if row['lval'] in factors and row['rval'] in factors:
            corr_matrix.loc[row['lval'], row['rval']] = row['Est. Std']
            corr_matrix.loc[row['rval'], row['lval']] = row['Est. Std']

    print("\n" + "=" * 90)
    print("FACTOR CORRELATIONS")
    print("=" * 90)
    print(corr_matrix.round(3).to_string())

    # Check for high correlations and Heywood cases
    off_diag = corr_matrix.values[np.triu_indices(len(factors), k=1)]
    high_corrs = []
    heywood_cases = []

    for i in range(len(factors)):
        for j in range(i + 1, len(factors)):
            val = corr_matrix.iloc[i, j]
            pair = f"{factors[i]}×{factors[j]}"
            if abs(val) > 0.85:
                high_corrs.append({'pair': pair, 'r': val})
                print(f"  ⚠️ HIGH CORRELATION: {pair} = {val:.3f}")
            if abs(val) > 1.0:
                heywood_cases.append({'pair': pair, 'r': val})
                print(f"  🚨 HEYWOOD CASE: {pair} = {val:.3f}")

    if not high_corrs:
        print("  ✓ No problematically high factor correlations detected")

    # ── Reliability ───────────────────────────────────────────────────────
    reliability_results = []
    for factor, items in CFA_STRUCTURE.items():
        item_data = df_cfa[items].dropna()
        alpha_result = pg.cronbach_alpha(item_data)
        alpha = alpha_result[0]

        # Factor loadings for this factor
        factor_loadings = loading_df[loading_df['Factor'] == factor]['Loading'].values

        # Composite Reliability
        sum_loadings = np.sum(factor_loadings)
        sum_error_var = np.sum(1 - factor_loadings**2)
        cr = sum_loadings**2 / (sum_loadings**2 + sum_error_var)

        # Average Variance Extracted
        ave = np.mean(factor_loadings**2)

        reliability_results.append({
            'Factor': factor,
            'N_Items': len(items),
            'Alpha': alpha,
            'CR': cr,
            'AVE': ave
        })

    reliability_df = pd.DataFrame(reliability_results)

    print("\n" + "=" * 90)
    print("RELIABILITY SUMMARY")
    print("=" * 90)
    print(reliability_df[['Factor', 'N_Items', 'Alpha', 'CR', 'AVE']].round(3).to_string(index=False))

    good_alpha = len(reliability_df[reliability_df['Alpha'] >= 0.70])
    good_cr = len(reliability_df[reliability_df['CR'] >= 0.70])
    good_ave = len(reliability_df[reliability_df['AVE'] >= 0.50])
    total = len(reliability_df)
    print(f"\n  Good α (≥.70): {good_alpha}/{total}")
    print(f"  Good CR (≥.70): {good_cr}/{total}")
    print(f"  Good AVE (≥.50): {good_ave}/{total}")

    # ── Dropped Items Evaluation ──────────────────────────────────────────
    print("\n" + "=" * 90)
    print("DROPPED ITEMS RELIABILITY EVALUATION")
    print("=" * 90)

    dropped_reliability = []
    for construct, items in DROPPED_ITEMS.items():
        if all(item in df_full.columns for item in items):
            item_data = df_full[items].dropna()
            alpha_result = pg.cronbach_alpha(item_data)
            alpha = alpha_result[0]
            inter_corr = item_data.corr().iloc[0, 1]
            status = '✓ Acceptable' if alpha >= 0.60 else '✗ Poor'
            dropped_reliability.append({
                'Construct': construct, 'Items': ', '.join(items),
                'N_Items': len(items), 'Alpha': alpha,
                'Inter_Item_r': inter_corr, 'Status': status
            })

    dropped_df = pd.DataFrame(dropped_reliability)
    print(dropped_df.to_string(index=False))

    # ── Export Results ────────────────────────────────────────────────────
    print("\n" + "=" * 90)
    print("EXPORTING RESULTS")
    print("=" * 90)

    loading_df.to_csv('tables/cfa_factor_loadings.csv', index=False)
    print("✓ tables/cfa_factor_loadings.csv")

    reliability_df.to_csv('tables/cfa_reliability.csv', index=False)
    print("✓ tables/cfa_reliability.csv")

    if len(CFA_STRUCTURE) >= 2:
        corr_matrix.to_csv('tables/cfa_factor_correlations.csv')
        print("✓ tables/cfa_factor_correlations.csv")

    cfa_summary = {
        'population': POPULATION,
        'n_observations': len(df_cfa),
        'n_factors': len(CFA_STRUCTURE),
        'n_items': sum(len(items) for items in CFA_STRUCTURE.values()),
        'factor_structure': CFA_STRUCTURE,
        'fit_indices': {
            'chi2': float(chi2) if chi2 else None,
            'df': int(df_val) if df_val else None,
            'CFI': float(cfi) if cfi else None,
            'TLI': float(tli) if tli else None,
            'RMSEA': float(rmsea) if rmsea else None,
            'SRMR': float(srmr_value) if srmr_value is not None else None,
            'AIC': float(aic) if aic else None,
            'BIC': float(bic) if bic else None
        },
        'factor_correlation_diagnostics': {
            'correlation_range': [float(min(off_diag)), float(max(off_diag))],
            'n_pairs_above_85': len(high_corrs),
            'n_heywood_cases': len(heywood_cases),
            'heywood_pairs': heywood_cases,
            'high_correlation_pairs': high_corrs
        } if len(CFA_STRUCTURE) >= 2 else None,
        'reliability': reliability_df[['Factor', 'Alpha', 'CR', 'AVE']].to_dict('records')
    }

    with open('tables/cfa_summary.json', 'w') as f:
        json.dump(cfa_summary, f, indent=2)
    print("✓ tables/cfa_summary.json")

    # ── Final Summary ─────────────────────────────────────────────────────
    print("\n" + "=" * 90)
    print("CFA ANALYSIS COMPLETE")
    print("=" * 90)
    print(f"  Sample: {POPULATION.title()} (N = {len(df_cfa)})")
    print(f"  Factors: {len(CFA_STRUCTURE)}, Items: {sum(len(i) for i in CFA_STRUCTURE.values())}")
    print(f"  CFI = {cfi:.3f}, TLI = {tli:.3f}, RMSEA = {rmsea:.3f}, SRMR = {srmr_value:.4f}")
    print(f"  Heywood cases: {len(heywood_cases)}")
    if heywood_cases:
        for h in heywood_cases:
            print(f"    {h['pair']} = {h['r']:.3f}")


if __name__ == '__main__':
    main()

File: airs_experiment/R/run_02_cfa.R
#!/usr/bin/env Rscript
# =============================================================================
# AIRS Phase 2: Confirmatory Factor Analysis (CFA) — R port
# Mirrors: python/run_02_cfa.py
#
# Validates EFA-derived factor structure using independent holdout sample.
# Method: lavaan CFA | Fit criteria: CFI/TLI >= 0.95, RMSEA <= 0.08, SRMR <= 0.08
# Outputs: tables/cfa_summary.json, tables/cfa_factor_loadings.csv,
#          tables/cfa_factor_correlations.csv, tables/cfa_reliability.csv
# =============================================================================

library(lavaan)
library(semTools)
library(psych)
library(jsonlite)

cat(paste(rep("=", 90), collapse = ""), "\n")
cat("PHASE 2: CONFIRMATORY FACTOR ANALYSIS (CFA)\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

dir.create("tables", showWarnings = FALSE)

# =============================================================================
# LOAD DATA
# =============================================================================

df_holdout <- read.csv("data/AIRS_holdout.csv", stringsAsFactors = FALSE)
cat(sprintf("\nLoaded holdout data: %d rows x %d columns\n", nrow(df_holdout), ncol(df_holdout)))

config <- fromJSON("data/experiment_config.json")
efa_summary <- fromJSON("tables/efa_summary.json")

# =============================================================================
# BUILD CFA STRUCTURE FROM EFA RESULTS
# =============================================================================

CONSTRUCT_NAME_MAP <- c(
  PE = "PerfExp", EE = "EffortExp", SI = "SocialInf", FC = "FacilCond",
  HM = "HedonicMot", PV = "PriceValue", HB = "Habit", TR = "AITrust"
)

DROPPED_ITEMS <- list(
  Voluntariness = c("VO1", "VO2"),
  Explainability = c("EX1", "EX2"),
  EthicalRisk = c("ER1", "ER2"),
  Anxiety = c("AX1", "AX2")
)

# Get validated structure from EFA
factor_struct <- efa_summary$validated_factor_structure
CFA_STRUCTURE <- list()
for (abbr in names(factor_struct)) {
  mapped_name <- CONSTRUCT_NAME_MAP[abbr]
  if (!is.na(mapped_name)) {
    CFA_STRUCTURE[[mapped_name]] <- factor_struct[[abbr]]
  }
}

# Build lavaan syntax
cfa_syntax_parts <- sapply(names(CFA_STRUCTURE), function(factor) {
  items <- CFA_STRUCTURE[[factor]]
  sprintf("  %s =~ %s", factor, paste(items, collapse = " + "))
})
cfa_syntax <- paste(cfa_syntax_parts, collapse = "\n")

cat("\nCFA Model Syntax:\n")
cat(cfa_syntax, "\n\n")

n_factors <- length(CFA_STRUCTURE)
n_items   <- sum(sapply(CFA_STRUCTURE, length))
cat(sprintf("CFA Structure: %d factors, %d items\n", n_factors, n_items))
cat(sprintf("Dropped: %s\n\n",
            paste(names(DROPPED_ITEMS), collapse = ", ")))

# Collect all CFA items
all_cfa_items <- unlist(CFA_STRUCTURE, use.names = FALSE)

# =============================================================================
# FIT CFA MODEL
# =============================================================================

fit_cfa <- cfa(cfa_syntax, data = df_holdout, estimator = "ML")
cat(sprintf("Converged: %s\n", lavInspect(fit_cfa, "converged")))

# Fit indices
fit_measures <- fitMeasures(fit_cfa, c("chisq", "df", "pvalue",
                                        "cfi", "tli", "rmsea", "rmsea.ci.lower", "rmsea.ci.upper",
                                        "srmr", "aic", "bic"))

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("CFA FIT INDICES (Holdout Sample)\n")
cat(paste(rep("=", 90), collapse = ""), "\n")
cat(sprintf("  Chi-square = %.3f (df = %d, p = %.4f)\n",
            fit_measures["chisq"], fit_measures["df"], fit_measures["pvalue"]))
cat(sprintf("  CFI  = %.4f  %s\n", fit_measures["cfi"],
            ifelse(fit_measures["cfi"] >= 0.95, "[EXCELLENT]", ifelse(fit_measures["cfi"] >= 0.90, "[ACCEPTABLE]", "[POOR]"))))
cat(sprintf("  TLI  = %.4f  %s\n", fit_measures["tli"],
            ifelse(fit_measures["tli"] >= 0.95, "[EXCELLENT]", ifelse(fit_measures["tli"] >= 0.90, "[ACCEPTABLE]", "[POOR]"))))
cat(sprintf("  RMSEA = %.4f [%.4f, %.4f]  %s\n",
            fit_measures["rmsea"], fit_measures["rmsea.ci.lower"], fit_measures["rmsea.ci.upper"],
            ifelse(fit_measures["rmsea"] <= 0.06, "[EXCELLENT]", ifelse(fit_measures["rmsea"] <= 0.08, "[ACCEPTABLE]", "[POOR]"))))
cat(sprintf("  SRMR = %.4f  %s\n", fit_measures["srmr"],
            ifelse(fit_measures["srmr"] <= 0.08, "[EXCELLENT]", "[POOR]")))
cat(sprintf("  AIC  = %.2f\n", fit_measures["aic"]))
cat(sprintf("  BIC  = %.2f\n", fit_measures["bic"]))

# =============================================================================
# FACTOR LOADINGS
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("STANDARDIZED FACTOR LOADINGS\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

std_solution <- standardizedSolution(fit_cfa)
loadings_df <- std_solution[std_solution$op == "=~",
                             c("lhs", "rhs", "est.std", "se", "z", "pvalue")]
colnames(loadings_df) <- c("Factor", "Item", "Loading_Std", "SE", "z", "p")

for (i in seq_len(nrow(loadings_df))) {
  sig <- ifelse(loadings_df$p[i] < 0.001, "***",
                ifelse(loadings_df$p[i] < 0.01, "**",
                       ifelse(loadings_df$p[i] < 0.05, "*", "")))
  cat(sprintf("  %s -> %s: %.4f (SE=%.4f, z=%.3f) %s\n",
              loadings_df$Factor[i], loadings_df$Item[i],
              loadings_df$Loading_Std[i], loadings_df$SE[i],
              loadings_df$z[i], sig))
}

write.csv(loadings_df, "tables/cfa_factor_loadings.csv", row.names = FALSE)
cat("\nExported: tables/cfa_factor_loadings.csv\n")

# =============================================================================
# FACTOR CORRELATIONS
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("FACTOR CORRELATIONS\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

cor_lv <- lavInspect(fit_cfa, "cor.lv")
cat("\n")
print(round(cor_lv, 3))

write.csv(round(cor_lv, 4), "tables/cfa_factor_correlations.csv")
cat("\nExported: tables/cfa_factor_correlations.csv\n")

# Discriminant validity check
fnames <- rownames(cor_lv)
cat("\nDiscriminant validity (|r| >= .85):\n")
any_flagged <- FALSE
for (i in 1:(length(fnames) - 1)) {
  for (j in (i + 1):length(fnames)) {
    r <- cor_lv[i, j]
    if (abs(r) >= 0.85) {
      cat(sprintf("  WARNING: %s x %s: r = %.3f\n", fnames[i], fnames[j], r))
      any_flagged <- TRUE
    }
  }
}
if (!any_flagged) cat("  None — all |r| < .85\n")

# =============================================================================
# RELIABILITY: Composite Reliability (CR) & AVE
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("RELIABILITY: CR & AVE\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

reliability_df <- data.frame(
  Factor = character(), N_Items = integer(),
  CR = numeric(), AVE = numeric(), Alpha = numeric(),
  CR_Status = character(), AVE_Status = character(),
  stringsAsFactors = FALSE
)

for (factor in names(CFA_STRUCTURE)) {
  items <- CFA_STRUCTURE[[factor]]
  # Get standardized loadings for this factor
  factor_loadings <- loadings_df$Loading_Std[loadings_df$Factor == factor]

  # CR = (sum of loadings)^2 / ((sum of loadings)^2 + sum of error variances)
  sum_lambda <- sum(factor_loadings)
  sum_error  <- sum(1 - factor_loadings^2)
  cr <- sum_lambda^2 / (sum_lambda^2 + sum_error)

  # AVE = mean of squared loadings
  ave <- mean(factor_loadings^2)

  # Cronbach's alpha
  alpha_val <- psych::alpha(df_holdout[, items], check.keys = FALSE)$total$raw_alpha

  reliability_df <- rbind(reliability_df, data.frame(
    Factor = factor, N_Items = length(items),
    CR = cr, AVE = ave, Alpha = alpha_val,
    CR_Status = ifelse(cr >= 0.70, "OK", "LOW"),
    AVE_Status = ifelse(ave >= 0.50, "OK", "LOW"),
    stringsAsFactors = FALSE
  ))

  cat(sprintf("  %s: CR=%.3f %s, AVE=%.3f %s, Alpha=%.3f\n",
              factor, cr, ifelse(cr >= 0.70, "[OK]", "[LOW]"),
              ave, ifelse(ave >= 0.50, "[OK]", "[LOW]"),
              alpha_val))
}

write.csv(reliability_df, "tables/cfa_reliability.csv", row.names = FALSE)
cat("\nExported: tables/cfa_reliability.csv\n")

# =============================================================================
# FORNELL-LARCKER CRITERION
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("FORNELL-LARCKER CRITERION\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

ave_vals <- setNames(reliability_df$AVE, reliability_df$Factor)
violations <- 0
for (i in 1:(length(fnames) - 1)) {
  for (j in (i + 1):length(fnames)) {
    r <- cor_lv[i, j]
    r2_shared <- r^2
    ave_i <- ave_vals[fnames[i]]
    ave_j <- ave_vals[fnames[j]]
    pass <- (ave_i > r2_shared) & (ave_j > r2_shared)
    if (!pass) {
      cat(sprintf("  FAIL: %s x %s: r=%.3f, r2=%.3f, AVE(%s)=%.3f, AVE(%s)=%.3f\n",
                  fnames[i], fnames[j], r, r2_shared, fnames[i], ave_i, fnames[j], ave_j))
      violations <- violations + 1
    }
  }
}
if (violations == 0) {
  cat("  All pairs pass Fornell-Larcker criterion\n")
} else {
  cat(sprintf("\n  Total violations: %d\n", violations))
}

# =============================================================================
# HTMT RATIO (Henseler et al., 2015)
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("HETEROTRAIT-MONOTRAIT RATIO (HTMT)\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

# Item-level correlation matrix on holdout data
all_items <- unlist(CFA_STRUCTURE, use.names = FALSE)
item_cor <- cor(df_holdout[, all_items], use = "pairwise.complete.obs")

htmt_results <- data.frame(
  Factor_i = character(), Factor_j = character(),
  HTMT = numeric(), Status = character(),
  stringsAsFactors = FALSE
)

factor_names <- names(CFA_STRUCTURE)
for (i in 1:(length(factor_names) - 1)) {
  for (j in (i + 1):length(factor_names)) {
    items_i <- CFA_STRUCTURE[[factor_names[i]]]
    items_j <- CFA_STRUCTURE[[factor_names[j]]]

    # Heterotrait-heteromethod: correlations between items of different factors
    het_cors <- abs(item_cor[items_i, items_j])
    mean_het <- mean(het_cors)

    # Monotrait-heteromethod: correlations within each factor
    if (length(items_i) > 1) {
      mono_i_cors <- abs(item_cor[items_i, items_i])
      mono_i <- mean(mono_i_cors[upper.tri(mono_i_cors)])
    } else {
      mono_i <- 1
    }
    if (length(items_j) > 1) {
      mono_j_cors <- abs(item_cor[items_j, items_j])
      mono_j <- mean(mono_j_cors[upper.tri(mono_j_cors)])
    } else {
      mono_j <- 1
    }

    htmt_val <- mean_het / sqrt(mono_i * mono_j)
    status <- ifelse(htmt_val < 0.85, "OK", ifelse(htmt_val < 0.90, "MARGINAL", "FAIL"))

    htmt_results <- rbind(htmt_results, data.frame(
      Factor_i = factor_names[i], Factor_j = factor_names[j],
      HTMT = htmt_val, Status = status,
      stringsAsFactors = FALSE
    ))

    if (htmt_val >= 0.85) {
      cat(sprintf("  %s: %s x %s: HTMT = %.3f\n", status, factor_names[i], factor_names[j], htmt_val))
    }
  }
}

htmt_ok <- sum(htmt_results$Status == "OK")
htmt_fail <- sum(htmt_results$Status != "OK")
cat(sprintf("\n  Summary: %d pairs OK, %d pairs flagged (HTMT >= .85)\n", htmt_ok, htmt_fail))
write.csv(htmt_results, "tables/cfa_htmt.csv", row.names = FALSE)
cat("  Exported: tables/cfa_htmt.csv\n")

# =============================================================================
# ALSO FIT ON FULL SAMPLE
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("CFA ON FULL SAMPLE (for comparison)\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

df_full <- read.csv("data/AIRS_clean.csv", stringsAsFactors = FALSE)
fit_cfa_full <- cfa(cfa_syntax, data = df_full, estimator = "ML")
fm_full <- fitMeasures(fit_cfa_full, c("chisq", "df", "pvalue",
                                        "cfi", "tli", "rmsea", "srmr", "aic", "bic"))
cat(sprintf("  N = %d\n", nrow(df_full)))
cat(sprintf("  CFI = %.4f, TLI = %.4f, RMSEA = %.4f, SRMR = %.4f\n",
            fm_full["cfi"], fm_full["tli"], fm_full["rmsea"], fm_full["srmr"]))

# =============================================================================
# EXPORT SUMMARY JSON
# =============================================================================

cfa_summary <- list(
  analysis_date = format(Sys.time(), "%Y-%m-%dT%H:%M:%S"),
  population = config$population,
  holdout_n = nrow(df_holdout),
  full_n = nrow(df_full),
  n_factors = n_factors,
  n_items = n_items,
  dropped_constructs = names(DROPPED_ITEMS),
  cfa_structure = CFA_STRUCTURE,
  fit_indices_holdout = as.list(round(fit_measures, 6)),
  fit_indices_full = as.list(round(fm_full, 6)),
  factor_loadings = lapply(seq_len(nrow(loadings_df)), function(i) {
    list(factor = loadings_df$Factor[i], item = loadings_df$Item[i],
         loading = round(loadings_df$Loading_Std[i], 4),
         p = loadings_df$p[i])
  }),
  reliability = lapply(seq_len(nrow(reliability_df)), function(i) {
    list(factor = reliability_df$Factor[i],
         CR = round(reliability_df$CR[i], 4),
         AVE = round(reliability_df$AVE[i], 4),
         Alpha = round(reliability_df$Alpha[i], 4))
  }),
  fornell_larcker_violations = violations,
  htmt = lapply(seq_len(nrow(htmt_results)), function(i) {
    list(factor_i = htmt_results$Factor_i[i], factor_j = htmt_results$Factor_j[i],
         htmt = round(htmt_results$HTMT[i], 4), status = htmt_results$Status[i])
  }),
  htmt_violations = htmt_fail,
  factor_correlations = round(cor_lv, 4)
)

write_json(cfa_summary, "tables/cfa_summary.json", pretty = TRUE, auto_unbox = TRUE)
cat("\nExported: tables/cfa_summary.json\n")

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("PHASE 2 COMPLETE\n")
cat(paste(rep("=", 90), collapse = ""), "\n")
